### PR TITLE
Fix coding style

### LINKED ISSUE #1275: Improve coding standards
Our [current ones](http://zotonic.com/docs/latest/developer-guide/contributing.html#coding-standards) are very limited.

Questions:
1. order of `-spec` and `%% @doc` function header
2. indent for function bodies
3. explain how to enable [Editconfig plugins](http://editorconfig.org/#download) in editors ([Emacs](https://github.com/editorconfig/editorconfig-emacs), [Vim](https://github.com/editorconfig/editorconfig-vim)), as I see al lot of (e.g.) trailing whitespace in the code, which causes diffs to be unnecessarily large

Does anyone know of linting tools that can catch CS violations?

### PR DESCRIPTION
**Pull request by ddeboer**
### Description

Fix #1275.

This is going to be a huge, fat, ugly pull request, but what the heck – let’s do it anyway. :wink:

### Checklist

- [x] no BC breaks
- [x] Re-enable Elvis

### CODE REVIEW

File: modules/mod_signal/mod_signal.erl [hunk 1]
```diff
@@ -182,11 +195,11 @@ handle_call({'connect', Signal, Slot}, _From, State) ->
...
-                  true ->
-                      {slot, Key, Slot};
-                  false ->
-                      [{slot, Key, Slot}, {tags, SignalType, SignalTags}]
-              end,
+        true ->
+            {slot, Key, Slot};
+        false ->
+            [{slot, Key, Slot}, {tags, SignalType, SignalTags}]
+    end,
```
Comment (mworrell) [inline modules/mod_signal/mod_signal.erl]: Actually I find these *left-aligned* case statements uglier and much less readable then the normal *aligned below the case* versions.

Comment (ddeboer) [inline modules/mod_signal/mod_signal.erl]: I can imagine. The problem with aligned below is that (depending on the preceding statement) they can jump all over the place and start very far to the right end leaving only a little line length.

**Comment by mworrell** [comment]
This will give quite some merge conflicts with my SSL branch.
Can you wait for that one?

Something else: maybe 132 columns?  The line printers we used in the 80s had already 132 columns... and it is still quite readable.

At least it is not 80 column punch cards.

File: modules/mod_signal/mod_signal.erl (hunk 1 above)
Comment (mworrell) [inline modules/mod_signal/mod_signal.erl]: If it is starting to jump too far then we that case statement should be lifted out and made into its own function :)

**Comment by ddeboer** [comment]
> Can you wait for that one?

Sure. Any idea as to what still needs to be done for the SSL PR? Just the docs?

> maybe 132 columns? The line printers we used in the 80s had already 132 columns... and it is still quite readable.

Let’s just follow common practice here as well. But Inaka specifies 80, which is (very) narrow, so I’m going with 110 for now.

**Comment by mworrell** [comment]
Did Elvis do this automatically, there are quite some changes that are unexpected (to say it nicely :) )

**Comment by mmzeeman** [comment]
BTW.

Erlang has automatic tools to tidy code automatically. This is an escript to use `erl_tidy` from the commandline.

```erlang
#!/usr/bin/env escript

main([Name]) ->
    erl_tidy:file(Name, config());
main(_) ->
    usage().

config() ->
    [{stdout, true},
     {keep_unused, true},
     %% erl_prettypr settings
     {paper, 132},
     {ribbon, 90}].

usage() ->
    io:put_chars("usage: erlfmt.escript <file>\n"),
    halt(1).
```

**Comment by ddeboer** [comment]
Thanks for the tip, @mmzeeman. I’ve been playing around with automatic tools, but some have some unwanted consequences, as noted by @mworrell.

**Comment by mmzeeman** [comment]
For example:

This `support/z_xml.erl`

```erlang
-module(z_xml).
-author("Marc Worrell <marc@worrell.nl").

%% interface functions
-export([
    escape/1
]).


%% @doc Escape a html text to valid a xml text so that it can be transported in XML.  Translates control characters to
%% spaces, except for TAB, CR and LF.
%% @spec escape(iolist()) -> iolist()
escape(undefined) ->
    undefined;
escape(<<>>) ->
    <<>>;
escape([]) ->
    [];
escape(L) when is_list(L) ->
    escape(list_to_binary(L));
escape(B) when is_binary(B) ->
    escape(B, <<>>).

    escape(<<>>, Acc) ->
        Acc;
    escape(<<$&, T/binary>>, Acc) ->
        escape(T, <<Acc/binary, "&#38;">>);
    escape(<<$<, T/binary>>, Acc) ->
        escape(T, <<Acc/binary, "&#60;">>);
    escape(<<$>, T/binary>>, Acc) ->
        escape(T, <<Acc/binary, "&#62;">>);
    escape(<<$", T/binary>>, Acc) ->
        escape(T, <<Acc/binary, "&#34;">>);
    escape(<<$', T/binary>>, Acc) ->
        escape(T, <<Acc/binary, "&#39;">>);
    escape(<<C, T/binary>>, Acc) when C == 9; C == 10; C == 13 ->
        escape(T, <<Acc/binary, C>>);
    escape(<<C, T/binary>>, Acc) when C < 32 ->
        escape(T, <<Acc/binary, 32>>);
    escape(<<C, T/binary>>, Acc) ->
        escape(T, <<Acc/binary, C>>).
```

Is turned into this:

```erlang
-module(z_xml).

-author("Marc Worrell <marc@worrell.nl").

%% interface functions
-export([escape/1]).

%% @doc Escape a html text to valid a xml text so that it can be transported in XML.  Translates control characters to
%% spaces, except for TAB, CR and LF.
%% @spec escape(iolist()) -> iolist()
escape(undefined) -> undefined;
escape(<<>>) -> <<>>;
escape([]) -> [];
escape(L) when is_list(L) -> escape(list_to_binary(L));
escape(B) when is_binary(B) -> escape(B, <<>>).

escape(<<>>, Acc) -> Acc;
escape(<<$&, T/binary>>, Acc) -> escape(T, <<Acc/binary, "&#38;">>);
escape(<<$<, T/binary>>, Acc) -> escape(T, <<Acc/binary, "&#60;">>);
escape(<<$>, T/binary>>, Acc) -> escape(T, <<Acc/binary, "&#62;">>);
escape(<<$", T/binary>>, Acc) -> escape(T, <<Acc/binary, "&#34;">>);
escape(<<$', T/binary>>, Acc) -> escape(T, <<Acc/binary, "&#39;">>);
escape(<<C, T/binary>>, Acc) when C == 9; C == 10; C == 13 ->
    escape(T, <<Acc/binary, C>>);
escape(<<C, T/binary>>, Acc) when C < 32 -> escape(T, <<Acc/binary, 32>>);
escape(<<C, T/binary>>, Acc) -> escape(T, <<Acc/binary, C>>).
```

**Comment by ddeboer** [comment]
Even that is a bit too aggressive: what’s wrong with functions on two lines?

**Comment by mworrell** [comment]
Functions on one line are quite common, then it is easier to see the pattern matching.

**Comment by ddeboer** [comment]
Unfortunately, that’s where things become too subjective, in my opinion. That’s why I’m relying on some [external standards](https://github.com/zotonic/zotonic/issues/1275#issuecomment-220820881).

**Comment by mworrell** [comment]
I think expressing intent and readability are *always* more important than some standard indentation.

**Comment by mworrell** [comment]
The example `z_xml` module @mmzeeman showed above is very clear after the tidy. We should check that one in :-)

**Comment by ddeboer** [comment]
> I think expressing intent and readability are always more important than some standard indentation.

True, but that’s no reason to not follow a standard. Guidelines such as Inaka’s are meant to improve readability by improving consistency. After all, readability in Zotonic’s code is (currently) hurt the most by the extremely inconsistent nature of the code (which I do understand, as the code has grown over 7 years). Inconsistency is a problem particularly for newcomers to the framework. 

I consider coming up with and agreeing on our own standards a waste of time: person A considers X more readable, person B prefers Y for clear intent etc.. Moreover, to gain momentum in the Erlang community, it makes sense to join some existing standards instead of coming up with our own.

> The example z_xml module @mmzeeman showed above is very clear after the tidy. We should check that one in :-)

If you like it, please do. But then also come up with a general rule (in this case, when to choose one-line over multi-line functions) and an automated way to test this to prevent us preferring our own subjective prejudices. Only then will we be able to have consistent code. 

For reference, this is what `erl_tidy` does to `m_predicate`:

```
-export([all/1, flush/1, for_subject/2, get/2, id_to_name/2, insert/2, is_predicate/2,
	 is_used/2, m_find_value/3, m_to_list/2, m_value/2, name_to_id/2, name_to_id_check/2,
	 object_category/2, objects/2, subject_category/2, subjects/2, update_noflush/4]).
```

Not so readable – in my humble opinion, course.

So,

1. Which rules in https://github.com/inaka/erlang_guidelines don’t we agree with?
2. Which rules should we add (e.g. one-liners)?
3. Which disagreements about the changes in this PR do we have? I think most of it is about indents (after `case` and `fun`).

**Comment by mworrell** [comment]
The one-line rule is:

 * A function has multiple (similar) guards.
 * The function body is small

Most of the "rules" are actually common sense.

But I like to make clear structure.
So:

```erlang
       X = #somerecordname{id=Foo,
                           attr=Bar,
                           ...},
```
(had to play with the spaces here, GitHub doesn't to wysiwyg for the indentation)

and not:

```erlang
       X = #somerecordname{id=Foo,
                 attr=Bar,
                 ...},
```
(you can't easily see which fields are set)

but ok:

```erlang
       X = #somerecordname{
                 id=Foo,
                 attr=Bar,
                 ...},
```

Also:

```erlang
case some_function(
     arg, arg,
     arg)
of
     guard -> ...
     ...
end
```

and not:

```erlang
case some_function(
     arg, arg,
     arg) of
     guard -> ...
     ...
end
```
(you can't easily scan the guards)

With a `fun()` I also prefer to have the `end` at the same indentation as the `fun`.
In that way it is immediately clear what the fun is.

Same for `case` (or `if`) and `end`.

Those are conceptual blocks, and it helps to be able to see them as a single block, without searching in the first non-indented line to see what kind of block it is.

**Comment by mworrell** [comment]
Having said all this, as long I can still read the code and it won't throw me off because of some funky structure (the case guards example above) then I am completely ok with the cleanup.

**Comment by mmzeeman** [comment]
@ddeboer IMHO the code we have right now is consistent enough. If you want to change things, please go ahead. But if what we really want to do is attract new people there might be more effective ways to do that.

1) Have a clear slogan/mission of what we are trying to do with Zotonic. Zotonic is very different from the average run of the mill web framework/cms. It amazes me every time when other developers tell me what they have to do to get a site with 200k unique visitors running reliably. This is a no-brainer with zotonic. My guess is that if we explain what we are trying to do and how we do it, this will attract new people.

2) Make it easy for outsiders to join. Never turn away anyone who wants to contribute.

**Comment by ddeboer** [comment]
> Most of the "rules" are actually common sense.

Yup, but still the Zotonic code doesn’t follow many of those, for instance the maximum nesting level – which definitely helps readability. 

I find the one-liner ‘rule’ a little too vague (but it doesn’t really matter, as this PR doesn’t change those), but I definitely agree with both records and cases. Do you see any examples in this PR where something else happened? That must’ve been unintentional. 

> But if what we really want to do is attract new people there might be more effective ways to do that.

Of course: this PR is not primarily geared towards that. Just to be clear here: the main goals of this PR are to:

1. Follow some simple conventions to improve code consistency and readability
2. Automate those checks (using Elvis) for new code changes so we don’t have to worry about it.

We had too many alerts when I previously enabled Elvis, so we have to make the code consistent first so we can re-enable the checks.

**Comment by mworrell** [comment]
There were some places where the record fields and functions ended up un-aligned.
If most work is done then I can just walk through this and correct these by hand (hopefully not triggering Elvis errors :) )

**Comment by mmzeeman** [comment]
I don't like the the Inaka rules which enforce style like the max level nesting and friends. This has nothing to do with consistent formatting.

Also, for the alignment, I use proportional fonts to edit code, so that would be very hard for me. Still waiting for elastic tab stops to become common.

**Comment by ddeboer** [comment]
This fixes some CS issues and will have Elvis check src/*.erl and src/models/*.erl on Travis for any violations. The other directories are yet untouched.

**Pull request merged**